### PR TITLE
 Retry Policy for Jmx Connection #700

### PR DESCRIPTION
**Pull request by SajidRiaz138**

### CODE REVIEW

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetryBackoffStrategy.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    public int getMaxAttempts()
+    {
+        return myRetryPolicyConfig.getMaxAttempts();
+    }
+
+    public long calculateDelay(final int attempt)
+    {
+        long baseDelay = myRetryPolicyConfig.getDelay();
+        long calculatedDelay = baseDelay * (attempt * 2L);
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetryBackoffStrategy.java:50]: Make a constant of the 2 so it is more clear why it is a 2

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetryBackoffStrategy.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+package com.ericsson.bss.cassandra.ecchronos.application.spring;
+
+import com.ericsson.bss.cassandra.ecchronos.application.config.connection.RetryPolicyConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class RetryBackoffStrategy
+{
+    private static final Logger LOG = LoggerFactory.getLogger(RetryBackoffStrategy.class);
+    private static final int MILLISECONDS_1000 = 1000;
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetryBackoffStrategy.java:24]: Name it like ONE_SECOND_IN_MS to follow the naming trail of other defaults

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        this.retryBackoffStrategy = new RetryBackoffStrategy(config.getConnectionConfig().getJmxConnection().getRetryPolicyConfig());
+    }
+
+    @PostConstruct
+    public void startScheduler()
+    {
+        long initialDelay = retryBackoffStrategy.getInitialDelay();
+        long fixedDelay = retryBackoffStrategy.getFixedDelay();
+
+        LOG.info("Starting RetrySchedulerService with initialDelay={} ms and fixedDelay={} ms", initialDelay, fixedDelay);
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java:93]: debug level

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+    @VisibleForTesting
+    void retryNodes()
+    {
+        LOG.debug("Retrying unavailable nodes");
+        List<Node> unavailableNodes = findUnavailableNodes();
+
+        if (unavailableNodes.isEmpty())
+        {
+            LOG.info("No unavailable nodes found.");
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java:106]: What does this mean "no unavailable nodes found" and why is it important to info log?

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+        LOG.info("Starting RetrySchedulerService with initialDelay={} ms and fixedDelay={} ms", initialDelay, fixedDelay);
+
+        myScheduler.scheduleWithFixedDelay(this::retryNodes, initialDelay, fixedDelay, TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    void retryNodes()
+    {
+        LOG.debug("Retrying unavailable nodes");
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java:101]: warn level

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                return; // Successfully reconnected, exit method
+            }
+        }
+        markNodeUnreachable(node, nodeId);
+    }
+
+    private boolean tryReconnectToNode(final Node node, final UUID nodeId, final int attempt)
+    {
+        long delayMillis = retryBackoffStrategy.calculateDelay(attempt);
+        LOG.info("Attempting to reconnect to node: {}, attempt: {}", nodeId, attempt);
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java:152]: warn level

File: application/src/main/resources/ecc.yml [hunk 7]
```diff
@@ -93,6 +93,26 @@ connection:
     ## The default provider will be used unless another is specified.
     ##
     provider: com.ericsson.bss.cassandra.ecchronos.application.providers.AgentJmxConnectionProvider
+    retryPolicy:
+      ## Max number of attempts ecChronos will try to connect with Cassandra.
+      maxAttempts: 5
+      ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count powered by two.
+      ## If the current attempt is 4 and the default delay is 5 seconds, so ((4(attempt) x 2) x 5(default delay)) = 40 seconds.
+      ## If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay.
+      delay: 5
```
Comment (jwaeab) [inline application/src/main/resources/ecc.yml:102]: perhaps this config should be like

```
delay:
  start: 5
  max: 30
  unit: seconds

```
As it is now it is confusing. If all the comments are skipped in the config file, i assume people would expect unit to be applied to all settings under retryPolicy. that is, initialDelay and fixedDelay as well

**Review by jwaeab — COMMENTED** [review]
Also:
* Test negatives; no config set, misspelling eetc
* Add test cases with other values; start delay longer than max, initial longer than fixed etc

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java (hunk 4 above)
Comment (SajidRiaz138) [reply application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java:106]: it was just for manual testing! forget to remove it:

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    public RetryPolicyConfig()
+    {
+    }
+
+    private static final int DEFAULT_MAX_ATTEMPTS = 5;
+    private static final long DEFAULT_DELAY_IN_MS = 5000;
+    private static final long DEFAULT_MAX_DELAY_IN_MS = 30000;
+    private static final long DEFAULT_INITIAL_DELAY_IN_MS = 86400000;
+    private static final long DEFAULT_FIXED_DELAY_IN_MS = 86400000;
+    private static final String DEFAULT_TIME_UNIT_IN_SECOND = "seconds";
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java:32]: DEFAULT_TIME_UNIT_SECONDS

File: application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java [hunk 9]
```diff
@@ -32,13 +31,18 @@
...
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertThrows;
 
 public class TestConfig
 {
     private static final String DEFAULT_AGENT_FILE_NAME = "all_set.yml";
+    private static final String NOTHING_SET_AGENT_FILE_NAME = "nothing_set.yml";
+    private static final String TIME_UNIT_IN_SECOND = "seconds";
```
Comment (jwaeab) [inline application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java:42]: TIME_UNIT_SECONDS

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/DistributedJmxConnection.java [hunk 10]
```diff
@@ -35,16 +39,28 @@ public DistributedJmxConnection()
...
     @Override
     protected Class<?>[] expectedConstructor()
     {
         return new Class<?>[] {
-                Supplier.class,
-                DistributedNativeConnectionProvider.class,
-                EccNodesSync.class
+                                Supplier.class,
+                                DistributedNativeConnectionProvider.class,
+                                EccNodesSync.class
```
Comment (VictorCavichioli) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/DistributedJmxConnection.java:63]: Remove extra space

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+    private static final int DEFAULT_MAX_ATTEMPTS = 5;
+    private static final long DEFAULT_DELAY_IN_MS = 5000;
+    private static final long DEFAULT_MAX_DELAY_IN_MS = 30000;
+    private static final long DEFAULT_INITIAL_DELAY_IN_MS = 86400000;
+    private static final long DEFAULT_FIXED_DELAY_IN_MS = 86400000;
+    private static final String DEFAULT_TIME_UNIT_IN_SECONDS = "seconds";
+    private RetryPolicyConfig.RetryDelay myRetryDelay = new RetryPolicyConfig.RetryDelay();
+    private RetryPolicyConfig.RetrySchedule myRetrySchedule = new RetryPolicyConfig.RetrySchedule();
+    @JsonProperty ("maxAttempts")
```
Comment (VictorCavichioli) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java:35]: line between line 34 and 35

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+        return myRetrySchedule;
+    }
+
+    @JsonProperty ("retrySchedule")
+    public void setRetrySchedule(final RetrySchedule retrySchedule)
+    {
+        myRetrySchedule = retrySchedule;
+    }
+
+    private static long convertToMillis(final Long value, final String unit)
```
Comment (VictorCavichioli) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java:77]:
```java
 private static long convertToMillis(final Long value, final TimeUnit unit)
    {
        return unit.toMillis(value);
    }
```

When you get the unit, you can do this

```java
@JsonProperty("unit")
        public final void setTimeUnit(final String timeUnit)
        {
            myTimeUnit = TimeUnit.valueOf(timeUnit.toUpperCase(Locale.US));
        }
```

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+    public RetryPolicyConfig()
+    {
+    }
+
+    private static final int DEFAULT_MAX_ATTEMPTS = 5;
+    private static final long DEFAULT_DELAY_IN_MS = 5000;
+    private static final long DEFAULT_MAX_DELAY_IN_MS = 30000;
+    private static final long DEFAULT_INITIAL_DELAY_IN_MS = 86400000;
+    private static final long DEFAULT_FIXED_DELAY_IN_MS = 86400000;
+    private static final String DEFAULT_TIME_UNIT_IN_SECONDS = "seconds";
```
Comment (VictorCavichioli) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java:32]: String --> TimeUnit
"seconds" --> TimeUnit.SECONDS

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+     */
+    public static final class RetryDelay
+    {
+        public RetryDelay()
+        {
+
+        }
+
+        @JsonProperty ("start")
+        private long myDelay = DEFAULT_DELAY_IN_MS;
```
Comment (VictorCavichioli) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java:101]: Line space between properties

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+        @JsonProperty ("max")
+        private long myMaxDelay = DEFAULT_MAX_DELAY_IN_MS;
+        @JsonProperty ("unit")
+        private String myUnit = DEFAULT_TIME_UNIT_IN_SECONDS;
```
Comment (VictorCavichioli) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java:105]: String --> TimeUnit

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+    public static final class RetrySchedule
+    {
+        public RetrySchedule()
+        {
+
+        }
+
+        @JsonProperty ("initialDelay")
+        private long myInitialDelay = DEFAULT_INITIAL_DELAY_IN_MS;
```
Comment (VictorCavichioli) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/RetryPolicyConfig.java:173]: Line between properties

**Review by VictorCavichioli — CHANGES_REQUESTED** [review]

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+ */
+
+@Service
+public final class RetrySchedulerService implements DisposableBean
+{
+
+    private static final Logger LOG = LoggerFactory.getLogger(RetrySchedulerService.class);
+    private static final String COLUMN_NODE_ID = "node_id";
+    private static final String COLUMN_NODE_STATUS = "node_status";
+    private static final int SCHEDULER_AWAIT_TERMINATION = 60;
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/RetrySchedulerService.java:69]: DEFAULT_ .... _IN_SECONDS

**Review by jwaeab — APPROVED** [review]

**Pull request merged**